python/lsst/pipe/base/graph/_versionDeserializers.py

```diff
@@ -0,0 +1,524 @@
+import networkx as nx
+import pickle
+import struct
+import uuid
+
+from collections import defaultdict
+from typing import TYPE_CHECKING
+
+from lsst.utils import doImport
+from lsst.pex.config import Config
+from lsst.daf.butler import DimensionUniverse, Quantum, SerializedDimensionRecord, DimensionRecord
+
+from .quantumNode import QuantumNode, SerializedQuantumNode
+from ..pipeline import TaskDef
+from ..pipelineTask import PipelineTask
+from ._implDetails import _DatasetTracker, DatasetTypeName
+
+
+if TYPE_CHECKING:
+    from .graph import QuantumGraph
+
+
+class StructSizeDescriptor:
+    """This is basically a class level property. It exists to report the size
+    (number of bytes) of whatever the formatter string is for a deserializer
+    """
+    def __get__(self, inst, owner) -> int:
+        return struct.calcsize(owner.FMT_STRING)
+
+
+@dataclass
+class DeserializerBase:
+    FMT_STRING: ClassVar[str] = NotImplemented
+    structSize: ClassVar[StructSizeDescriptor]
+
+    preambleSize: int
+    sizeBytes: bytes
+
+    def __init_subclass__(cls) -> None:
+        # Veriy that each subclass has implemented a format string that
+        # describes how sizing info is packed into bytes
+        if cls.FMT_STRING is NotImplemented:
+            raise NotImplementedError("FMT_STRING must be implemented on all subclasses")
+        # attach the size decriptor
+        cls.structSize = StructSizeDescriptor()
+        super().__init_subclass__()
+
+    def unpackHeader(self, rawHeader: bytes) -> Optional[str]:
+        raise NotImplementedError
+
+    @property
+    def headerSize(self) -> int:
+        """Returns the number of bytes from the beginning of the file to the
+        end of the metadata.
+        """
+        raise NotImplementedError("Base class does not implement this method")
+
+    def readHeaderInfo(self, rawHeader: bytes) -> SimpleNamespace:
+        """Parse the supplied raw bytes into the header information and
+        byte ranges of specific TaskDefs and QuantumNodes
+
+        rawheader : bytes
+            The bytes that are to be parsed into the header information. These
+            are the bytes after the preamble and structsize number of bytes
+            and before the headerSize bytes
+        """
+        raise NotImplementedError("Base class does not implement this method")
+
+    def constructGraph(self, nodes: set[uuid.UUID], _readBytes: Callable,
+                       universe: DimensionUniverse) -> QuantumGraph:
+        raise NotImplementedError("Base class does not implement this method")
+
+    def description(self) -> str:
+        """Return the description of the serialized data format
+        """
+        raise NotImplementedError("Base class does not implement this method")
+
+
+Version1Description = """
+The save file starts with the first few bytes corresponding to the magic bytes
+in the QuantumGraph: `qgraph4\xf6\xe8\xa9`.
+
+The next few bytes are 2 big endian unsigned long longs.
```

Comment:
That is a good point, and it is what the code is doing. I was just using the terminology from the python struct documentation, and not really thinking about the fact that (in theory) it could differ.